kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py

```diff
@@ -0,0 +1,260 @@
+        tracker_pose_pub: publish tracking obstacles with PoseArray, for rviz visualization
+    '''
+
+    def __init__(self):
+        '''initialize attributes and setup subscriber and publisher'''
+
+        super().__init__('kf_hungarian_node')
+        self.declare_parameters(
+            namespace='',
+            parameters=[
+                ('global_frame', "camera_link"),
+                ('a_noise', [2., 2., 0.5]),
+                ('top_down', False),
+                ('death_threshold', 3),
+                ('measurementNoiseCov', [1., 1., 1.]),
+                ('errorCovPost', [1., 1., 1., 10., 10., 10.]),
+                ('vel_filter', [0.1, 2.0]),
+                ('height_filter', [-2.0, 2.0]),
+                ('cost_filter', 1.0)
+            ])
+        self.global_frame = self.get_parameter("global_frame")._value
+        self.death_threshold = self.get_parameter("death_threshold")._value
+        self.measurementNoiseCov = self.get_parameter("measurementNoiseCov")._value
+        self.errorCovPost = self.get_parameter("errorCovPost")._value
+        self.a_noise = self.get_parameter("a_noise")._value
+        self.vel_filter = self.get_parameter("vel_filter")._value
+        self.height_filter = self.get_parameter("height_filter")._value
+        self.top_down = self.get_parameter("top_down")._value
+        self.cost_filter = self.get_parameter("cost_filter")._value
+
+        self.obstacle_list = []
+        self.max_id = 0
+        self.sec = 0
+        self.nanosec = 0
+
+        # subscribe to detector 
+        self.detection_sub = self.create_subscription(
+            ObstacleArray,
+            'detection',
+            self.callback,
+            10)
+
+        # publisher for tracking result
+        self.tracker_obstacle_pub = self.create_publisher(ObstacleArray, 'tracking', 10)
+        self.tracker_marker_pub = self.create_publisher(MarkerArray, 'marker', 10)
+
+        # setup tf related
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+    def callback(self, msg):
+        '''callback function for detection result'''
+
+        # update delta time
+        dt = (msg.header.stamp.sec - self.sec) + (msg.header.stamp.nanosec - self.nanosec) / 1e9
+        self.sec = msg.header.stamp.sec
+        self.nanosec = msg.header.stamp.nanosec
+
+        # get detection
+        detections = msg.obstacles
+        num_of_detect = len(detections)
+        num_of_obstacle = len(self.obstacle_list)
+
+        # transform to global frame
+        if self.global_frame is not None:
+            try:
+                trans = self.tf_buffer.lookup_transform(self.global_frame, msg.header.frame_id, rclpy.time.Time())
+            except LookupException:
+                self.get_logger().info('fail to get tf from {} to {}'.format(msg.header.frame_id, self.global_frame))
+                return
+            else:
```

Comment:
This `else` doesn't line up with any if. I'm not sure this builds?

Reply:
You can remove the `else` as long as this is still under the `if self.global_frame is not None:`. Then if it needs to transform, it will, and if it fails it returns, but otherwise it'll update the messages.

Reply:
We can do `try...except...else...` in python

Reply:
Ah ok, that's bazaar to me, but ok. I don't think that `else` is needed then, is it? If the exception is thrown, it exits. So in all other cases, we want to run this.

Reply:
Yes, we can also go without the `else`